solr/core/src/java/org/apache/solr/jersey/SolrJerseyResponse.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.jersey;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Base response-body POJO to be used by Jersey resources.
+ *
+ * <p>Contains fields common to all Solr API responses, particularly the 'responseHeader' and
+ * 'error' fields.
+ */
+public class SolrJerseyResponse implements JacksonReflectMapWriter {
+
+  @JsonProperty("responseHeader")
+  public ResponseHeader responseHeader = new ResponseHeader();
+
+  @JsonProperty("error")
+  public ErrorInfo error;
+
+  public static class ResponseHeader implements JacksonReflectMapWriter {
```

Comment:
Rigidity.  I've seen and probably written plugins that add to the header as a map of whatever info you want from, say, a SearchComponent.  Will JAX-RS prevent such plugins from doing so?

Reply:
Oh, no, not at all.  We've got a few options to handle cases like that:

1. Individual APIs (e.g. `/select`) can still choose to return `NamedList<Object>` or other loosely typed objects instead of the "stronger" SolrJerseyResponse.  I think we should still strive to be strongly typed where-ever possible, but nothing about JAX-RS _requires_ this on any given API.  Loose-typing might be best for `/select`, which as you point out can have any number of custom SearchComponents that modify the response.
2. Alternatively, there are ways to get some of this flexibility in a strongly-typed object.  e.g. [Jackson's support for "additional properties"](https://stackoverflow.com/questions/39156293/jackson-deserialize-extra-fields-as-map).  It might take some work to expand a solution of this sort out to xml, javabin, etc. but it seems do-able if this is an approach we like.

Reply:
Just to re-iterate my support for strong-typing, I'd say a huge aspect of the benefits / point of this big PR is for us to get strong typing!  A major win.

Reply:
Oh for sure - very much looking forward to seeing the strong-typing start to cut down on the NamedList/Map types that just absolutely abound across the codebase.

How do you want to go forward on the rigidity issue?  Is that something you'd be fine deferring to a follow-up PR, or would you like to see that fixed here?  And either way, do you have a preference between the two potential solutions above?

My preference would be to go the Jackson "additional properties" route on `ResponseHeader` - that'd allow us to at least keep strong typing for the few header entries we know about.

Reply:
Definitely "additional properties" route.

Reply:
I just added some "additional properties" support, making use of the JsonAnyGetter/JsonAnySetter annotation pair described in the link we discussed above.  Jackson is able to use those annotations natively.

Also needed a few changes to `Utils.reflectWrite` to actually get this working for XML, javabin, etc.  Which isn't ideal, but is a reality of the way our MapWriter/ResponseWriter stuff works today.  We already depend on Jackson on the server-side, and it supports other formats - I wonder if we couldn't just use it for XML, etc. at some point in the future.  But that's of course a large and very different task.  Just thinking aloud here I guess...